cmd/ncproxy/hcn.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/Microsoft/hcsshim/hcn"
+	ncproxygrpc "github.com/Microsoft/hcsshim/pkg/ncproxy/ncproxygrpc/v1"
+	"github.com/pkg/errors"
+)
+
+func modifyEndpoint(ctx context.Context, id string, policies []hcn.EndpointPolicy, requestType hcn.RequestType) error {
+	endpointRequest := hcn.PolicyEndpointRequest{
+		Policies: policies,
+	}
+
+	settingsJSON, err := json.Marshal(endpointRequest)
+	if err != nil {
+		return err
+	}
+
+	requestMessage := &hcn.ModifyEndpointSettingRequest{
+		ResourceType: hcn.EndpointResourceTypePolicy,
+		RequestType:  requestType,
+		Settings:     settingsJSON,
+	}
+
+	return hcn.ModifyEndpointSettings(id, requestMessage)
+}
+
+func parseEndpointPolicies(policies []hcn.EndpointPolicy) (*ncproxygrpc.HcnEndpointPolicies, error) {
+	results := &ncproxygrpc.HcnEndpointPolicies{}
+	for _, policy := range policies {
+		switch policy.Type {
+		case hcn.PortMapping:
+			portMapSettings := &hcn.PortnameEndpointPolicySetting{}
+			if err := json.Unmarshal(policy.Settings, portMapSettings); err != nil {
+				return nil, err
+			}
+			results.PortnamePolicySetting = &ncproxygrpc.PortNameEndpointPolicySetting{
+				PortName: portMapSettings.Name,
+			}
+		case hcn.IOV:
+			iovSettings := &hcn.IovPolicySetting{}
+			if err := json.Unmarshal(policy.Settings, iovSettings); err != nil {
+				return nil, err
+			}
+			results.IovPolicySettings = &ncproxygrpc.IovEndpointPolicySetting{
+				IovOffloadWeight:    iovSettings.IovOffloadWeight,
+				QueuePairsRequested: iovSettings.QueuePairsRequested,
+				InterruptModeration: iovSettings.InterruptModeration,
+			}
+		}
+	}
+	return results, nil
+}
+
+func constructEndpointPolicies(req *ncproxygrpc.HcnEndpointPolicies) ([]hcn.EndpointPolicy, error) {
+	if req == nil {
```

Comment:
we don't check the req anywhere else. Is there a good reason to do it only here?

Reply:
This specifically checks if the policies field in the larger request is nil. I can move this check to the calling function though since that would be more consistent.

Reply:
Done :)